### PR TITLE
[RLlib] APPO (new API stack) enhancements vol 04: Smaller cleanups and preparations.

### PR DESCRIPTION
**Pull request by sven1977**
APPO (new API stack) enhancements vol 04: Smaller cleanups and preparations.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/torch/torch_learner.py [hunk 1]
```diff
@@ -147,23 +147,7 @@ def _uncompiled_update(
...
-        for key in off_policyness.keys():
-            mid = key[0]
-            if Columns.LOSS_MASK not in batch[mid]:
-                off_policyness[key] = torch.mean(off_policyness[key])
-            else:
-                mask = batch[mid][Columns.LOSS_MASK]
-                num_valid = torch.sum(mask)
-                off_policyness[key] = torch.sum(off_policyness[key][mask]) / num_valid
-        self.metrics.log_dict(off_policyness, window=1)
+        self._compute_off_policyness(batch)
```
Comment (simonsays1980) [inline rllib/core/learner/torch/torch_learner.py:150]: Nice!

**Review by simonsays1980 — APPROVED** [review]
LGTM. Some questions regarding metrics.

File: rllib/algorithms/impala/impala.py [hunk 2]
```diff
@@ -168,6 +162,9 @@ def __init__(self, algo_class=None):
         # __sphinx_doc_end__
         # fmt: on
 
+        # IMPALA takes care of its own EnvRunner (weights, connector, metrics) synching.
+        self._dont_auto_sync_env_runner_states = True
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/impala/impala.py:172]: This is global for RLlib (generic Algorithm logic), but thus far, only IMPALA and APPO use this setting. I would like to make this the standard soon for all algos ...

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -3215,7 +3216,17 @@ def _run_one_training_iteration(self) -> Tuple[ResultDict, "TrainIterCtx"]:
                             "one single result dict per training iteration."
                         )
 
-        # Only here, reduce the results into a single result dict.
+                    # TODO (sven): Resolve this metric through log_time's future
+                    #  ability to compute throughput.
+                    self.metrics.log_value(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm.py:3221]: Great question. For PPO, this is not applicable. Each loss reported is the only loss for that iteration. Or do you mean SGD minibatch losses?? Those are lost for now, since all losses are always logged with `window=1`.

File: rllib/core/learner/torch/torch_learner.py [hunk 4]
```diff
@@ -630,6 +614,25 @@ def _map_module_to_device(self, module: MultiRLModule) -> None:
                 if isinstance(module[key], torch.nn.Module):
                     module[key].to(self._device)
 
+    def _compute_off_policyness(self, batch):
+        # Log off-policy'ness of this batch wrt the current weights.
+        off_policyness = {
+            (mid, DIFF_NUM_GRAD_UPDATES_VS_SAMPLER_POLICY): (
+                (self._weights_seq_no - module_batch[WEIGHTS_SEQ_NO]).float()
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py:621]: not sure I understand the question.

We measure here the difference between a) number of updates the Learner has done thus far on the model vs b) number of updates the samplers' policies (on average) have undergone thus far. Since the samplers are always behind, this delta is always some small positive number. For PPO, it should actually be zero, for APPO/IMPALA somewhat < 100, etc..

File: rllib/algorithms/algorithm_config.py [hunk 5]
```diff
@@ -1798,15 +1803,16 @@ def env_runners(
...
+                details.
                 Tuning these values is important when running experiments with
                 large sample batches, where there is the risk that the object store may
                 fill up, causing spilling of objects to disk. This can cause any
                 asynchronous requests to become very slow, making your experiment run
                 slowly as well. You can inspect the object store during your experiment
-                via a call to Ray memory on your head node, and by using the Ray
+                via a call to `ray memory` on your head node, and by using the Ray
                 dashboard. If you're seeing that the object store is filling up,
                 turn down the number of remote requests in flight or enable compression.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:1817]: Added to next PR in the pipeline ...

File: rllib/algorithms/algorithm_config.py [hunk 6]
```diff
@@ -2135,6 +2142,10 @@ def learners(
                 an index into the available
                 CUDA devices. For example if `os.environ["CUDA_VISIBLE_DEVICES"] = "1"`
                 and `local_gpu_idx=0`, RLlib uses the GPU with ID=1 on the node.
+            max_requests_in_flight_per_learner: Max number of in-flight requests
+                to each Learner (actor)). See the
+                `ray.rllib.utils.actor_manager.FaultTolerantActorManager` class for more
+                details.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:2148]: Done on follow-up PR ...
